homeassistant/components/agent_dvr/config_flow.py

```diff
@@ -0,0 +1,118 @@
+from .const import DOMAIN, SERVER_URL  # pylint:disable=unused-import
+from .helpers import generate_url
+
+DEFAULT_PORT = 8090
+_LOGGER = logging.getLogger(__name__)
+
+
+class AgentFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle an Agent config flow."""
+
+    async def async_step_user(self, info=None):
+        """Handle an Agent config flow."""
+        errors = {}
+
+        if info is not None:
+            host = info[CONF_HOST]
+            port = info[CONF_PORT]
+
+            server_origin = generate_url(host, port)
+            agent_client = Agent(server_origin, async_get_clientsession(self.hass))
+
+            await agent_client.update()
+            await agent_client.close()
+
+            if agent_client.is_available:
+                await self.async_set_unique_id(agent_client.unique)
+
+                self._abort_if_unique_id_configured(
+                    updates={
+                        CONF_HOST: info[CONF_HOST],
+                        CONF_PORT: info[CONF_PORT],
+                        SERVER_URL: server_origin,
+                    }
+                )
+
+                self.device_config = {
+                    CONF_HOST: host,
+                    CONF_PORT: port,
+                    SERVER_URL: server_origin,
+                }
+
+                return await self._create_entry(agent_client.name)
+
+            errors["base"] = "device_unavailable"
+
+        data = self.discovery_schema or {
+            vol.Required(CONF_HOST): str,
+            vol.Required(CONF_PORT, default=DEFAULT_PORT): int,
+        }
+
+        return self.async_show_form(
+            step_id="user",
+            description_placeholders=self.device_config,
+            data_schema=vol.Schema(data),
+            errors=errors,
+        )
+
+    def __init__(self):
+        """Initialize the Agent config flow."""
+        self.device_config = {}
+        self.discovery_schema = {}
+
+    async def _create_entry(self, server_name):
+        """Create entry for device."""
+        return self.async_create_entry(title=server_name, data=self.device_config)
+
+    async def async_step_zeroconf(self, discovery_info=None):
+        """Prepare configuration for a discovered Agent device."""
+
+        if discovery_info is None:
+            return self.async_abort(reason="connection_error")
+
+        if discovery_info[CONF_HOST].startswith("169.254"):
+            return self.async_abort(reason="link_local_address")
+
+        host = discovery_info[CONF_HOST]
+        port = discovery_info[CONF_PORT]
+
+        server_origin = generate_url(host, port)
+        agent_client = Agent(server_origin, async_get_clientsession(self.hass))
+        await agent_client.update()
+        await agent_client.close()
+
+        if not agent_client.is_available:
+            return self.async_abort(reason="not_agent_device")
+
+        await self.async_set_unique_id(agent_client.unique)
+
+        self._abort_if_unique_id_configured(
```

Comment:
It looks like you have enough to ask them if they want to link the device without sending them though the user step.  See `harmony/config_flow.py` for an example.

Reply:
Can we add this to a future PR? I don't currently have a way of end to end testing any code changes to this part of the script.

Reply:
Absolutely, please remove the `zeroconf` step for now and add it back in a future pr when you are able to test